geomstats/geometry/product_manifold.py

```diff
@@ -14,101 +14,288 @@
+        trailing_shape = manifold.shape
+        new_shape = leading_shape + trailing_shape
+        return gs.reshape(argument, new_shape)
+
+    def _iterate_over_factors(self, func, args):
+        """Apply a function to each factor of the product.
+
+        If default_point_type is 'vector' then the vector must be split up into
+        sub-vectors for each component, and then those sub-vectors should be broadcast
+        into the correct shape for the individual manifold. These are then passed to the
+        function along with any non-array arguments.
+
+        However, if default_point_type is 'matrix', then we have an array of shape
+        (..., n_manifolds, dim_each). In this case we simply split into n_manifolds
+        arrays, each of shape (dim_each,).
+
+        The returned value will be array-like with all outputs contained in the
+        trailing dimension if default_point_type is vector, or in the two trailing
+        dimensions if default_point_type is matrix.
+
+        Parameters
+        ----------
+        func : str
+            The name of a method which is defined for each factor of the product
+            The method returns an array of shape (..., k)
+        args : dict
+            Dict of arguments.
+            Array-type arguments must be of type (..., shape)
+            Other arguments are passed to each factor unchanged
+
+        Returns
+        -------
+        out : array-like, shape = [..., {n_manifolds*k, (n_manifolds, k)}]
+        """
+        arguments, numerical_args, leading_dimensions = \
+            self._validate_args_for_iteration(args)
+
+        if self.default_point_type == "vector":
+            args_list = [
+                {
+                    key: self._reshape_trailing(arguments[key][j], self.factors[j])
+                    for key in arguments
+                }
+                for j in range(len(self.factors))
+            ]
+        elif self.default_point_type == "matrix":
+            args_list = [
+                {key: arguments[key][..., j, :] for key in arguments}
+                for j in range(len(self.factors))
+            ]
+
+        out = [self._get_method(
+            self.factors[i], func, {**args_list[i], **numerical_args}
         )
+            for i in range(len(self.factors))]
+
+        if self.default_point_type == 'vector':
+            # The individual factors might have matrix type, so their responses must be
+            # flattened before proceeding. Since it is not straightforward to see which
+            # type has been returned, we are going to have to compare the shape of the
+            # return to the leading shape of all arguments broadcast against each otehr
+            for response in out:
+                if (gs.is_array(response) and
+                        len(response.shape) > len(leading_dimensions) + 1):
+                    response.reshape(response.shape[:-2] + (-1,))
+            out = gs.concatenate(out, axis=-1)
```

Comment:
This part of the code fails for this simple case (the problem is in the concatenate), but changing it may not be trivial, as it may break other cases:

```python
import geomstats.backend as gs

from geomstats.geometry.euclidean import Euclidean
from geomstats.geometry.product_manifold import ProductManifold

factors = [Euclidean(3), Euclidean(2)]
space = ProductManifold(factors)

points = space.random_point(2)  # works

space.belongs(points)  # outputs True. Should output gs.array([True, True])
```

Another example is `space.is_tangent` with n `tangent_vec` and n `base_point`.